### PR TITLE
Test deriving errors

### PR DESCRIPTION
**Pull request by NathanReb**
This adds the boiler plate to test error cases along with a couple example test cases.

I'll add more cases later!

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+The successful cases are splitted between `test/deriver/test_default.ml` and
+`test/deriver/test_factory.ml` depending on which part of the plugin they correspond.  You can update
+those with new test cases and then run `dune runtest`. It will show the diff with the generated code
+for your new case, if you're happy with the result you can update the expected results by running
+`dune promote`.
+
+We also test the PPX errors to make sure they are properly triggered and located. There's one file
+per error case tested in `test/deriver/errors`. If you want to add a test case there you should:
+1. run `touch test/deriver/errors/factory_<test_case_name>.{ml,expected}` to add the new empty test
```
Comment (leamingrad) [inline CONTRIBUTING.md:38]: `deriver` -> `derivers`

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+They are run by comparing the output of `ppx_factory` as a standalone binary against `.expected`
+files, both for successful and unsuccessful deriving.
+
+The successful cases are splitted between `test/deriver/test_default.ml` and
+`test/deriver/test_factory.ml` depending on which part of the plugin they correspond.  You can update
+those with new test cases and then run `dune runtest`. It will show the diff with the generated code
+for your new case, if you're happy with the result you can update the expected results by running
+`dune promote`.
+
+We also test the PPX errors to make sure they are properly triggered and located. There's one file
```
Comment (leamingrad) [inline CONTRIBUTING.md:36]: This documentation is generally really nice :+1:

File: test/deriver/errors/default_str_no_manifest.ml [hunk 3]
```diff
@@ -0,0 +1 @@
+type t [@@deriving default]
```
Comment (leamingrad) [inline test/deriver/errors/default_str_no_manifest.ml]: Minor: Should this be on two lines so it looks more like our code?

**Review by leamingrad — COMMENTED** [review]
This PR is great - once it is in I'll steal most of it for ppx_enum.

High-level question: should we pull this function into a library for other ppx rewriters to use? (or push it into ppxlib)

File: test/deriver/errors/default_str_no_manifest.ml (hunk 3 above)
Comment (NathanReb) [reply test/deriver/errors/default_str_no_manifest.ml]: Yep I'll change that!

File: CONTRIBUTING.md (hunk 1 above)
Comment (NathanReb) [reply CONTRIBUTING.md:38]: In the path you mean ? The folder is indeed called `deriver`

Comment (NathanReb) [reply CONTRIBUTING.md:38]: ouh nevermind the diff wasn't showing properly!

**Comment by NathanReb** [comment]
The pattern is indeed gonna repeat a lot, I actually just copied it from what I did for `ppx_yojson` which I copied from what we came up with after dicussing this with @emillon.

I really like this setup and I guess it would be nice to be able to avoid repeating all that boiler plate and polish it a bit.

For instance I copied `pp.ml` to have the error test cases seperated from the rest in a folder to avoid polluting the main test one.
There's also the unnecessary library stanzas that I produce to make sure we will awlays have a `.merlin` there.
The `bash` action could be replaced by a `run?` if we contribute it to dune to make this more portable.

The soon to come `dune` plugins might help extracting all that into a reusable tool/lib/plugin.

It's probably a bit early now but once it has been battle-tested I'd gladly make it available to ppx authors!

**Review by leamingrad — APPROVED** [review]

**Pull request merged**